### ISSUE TITLE
equinox.nn.MLP with jax.experimental.jet.jet

### ISSUE THREAD

**Issue by A-Alaa** [patrick-kidger/equinox#1051]
Many thanks for this great library!

Using equinox.nn.MLP in `jax.experimental.jet.jet` transformation raises the `UnexpectedTracerError`. Using other equinox modules can work well, e.g. `equinox.nn.Linear`.

```python
import jax.numpy as jnp
import equinox as eqx
import jax.random as jr
from jax.experimental.jet import jet

@eqx.filter_jit
def f(x):
    return jnp.sin(x)

g = eqx.nn.MLP(1, 1, 10, 2, key=jr.PRNGKey(0))
h = eqx.nn.Linear(1, 1, key=jr.PRNGKey(1))

# Using f or h works. Using g crashes.
(y0, [*yns]) = jet(g, (jnp.array([0.0]),), ((jnp.ones(1, dtype=jnp.float32), ),))
```

The long stacktrace points to the following line in `equinox.nn.MLP.__call__`:
```python
x = filter_vmap(lambda a, b: a(b))(layer_activation, x)
```

**Comment by A-Alaa** [comment]
The following example make it more likely a jax issue.

```python
import jax.numpy as jnp
import jax.nn as jnn
import equinox as eqx
import jax.random as jr
from jax.experimental.jet import jet

# using `jnn.sigmoid` activation here instead of the default `jnn.relu` removes the error.
g = eqx.nn.MLP(1, 1, 10, 2, key=jr.PRNGKey(0), activation=jnn.sigmoid)

(y0, [*yns]) = jet(g, (jnp.array([0.0]),), ((jnp.ones(1, dtype=jnp.float32), ),))
```

So using sigmoid activation instead of the default relu removes the error. Any hint for a workaround to still use `jnn.relu`?

**Comment by A-Alaa** [comment]
It is purely jax issue, picked up [here](https://github.com/jax-ml/jax/issues/30352):
